### DISCUSSION TITLE
Custom converters - virtual attribute and time passed since last report

### DISCUSSION THREAD

**Discussion by Mr-Groch** [Koenkk/zigbee2mqtt#7048] [Devices]
Hi!

I want to create custom zigbee device - it will report some numeric value, but I don't want to expose that value - rather I want to make some mathematical operations and expose result. In those operations I need to have time passed since last report. It's that possible? Can I find example converters with such logic as reference?

**Comment by Koenkk** [comment]
Example can be found here:
- https://github.com/Koenkk/zigbee-herdsman-converters/blob/e3902f144238fa5a00027c02db33c75e5ad92c12/converters/fromZigbee.js#L4295
- https://github.com/Koenkk/zigbee-herdsman-converters/blob/e3902f144238fa5a00027c02db33c75e5ad92c12/converters/fromZigbee.js#L4308

**Reply by Mr-Groch** [reply]
Thank you! That is what I'm looking for! globalStore is the answer. So I can achieve this with external converter:

```yaml
external_converters:
  - ptvo.counter.js
```

And in that js file:

```js
...
const fzLocal = {
    counter: {
        cluster: 'genAnalogInput',
        type: ['attributeReport', 'readResponse'],
        convert: (model, msg, publish, options, meta) => {
            const payload = {};
...
                if (!globalStore.hasValue(msg.endpoint, 'counter')) {
                    globalStore.putValue(msg.endpoint, 'counter', {counter: val, since: Date.now()});
                }
                const last = globalStore.getValue(msg.endpoint, 'counter');
... // do something and add result to payload (device state attributes)
                globalStore.putValue(msg.endpoint, 'counter', {counter: val, since: Date.now()});
...
            return payload;
        },
    },
};

const device = {
...
    fromZigbee: [fz.ignore_basic_report, fzLocal.counter],
...
};

module.exports = device;
```
